rust/defi/src/defi_dapp/exchange.rs

```diff
@@ -176,83 +159,69 @@ impl Exchange {
-                    if let Some(b) = self.orders.get(&m.1) {
-                        // Check if some orders can be completed in their entirety.
-                        if b.from_amount >= a.to_amount {
-                            a_to_amount = a.to_amount;
-                        }
-                        if a.from_amount >= b.to_amount {
-                            b_to_amount = b.to_amount;
-                        }
-                        // Check if some orders can be completed partially.
-                        if a_to_amount == 0 && b_to_amount > 0 {
-                            a_to_amount = b.from_amount;
-                            // Verify that we can complete the partial order with natural number tokens remaining.
-                            if ((BigUint::from(a_to_amount) * BigUint::from(a.from_amount))
-                                % BigUint::from(a.to_amount))
-                                != BigUint::from(0u32)
-                            {
-                                continue;
-                            }
-                        }
-                        if b_to_amount == 0 && a_to_amount > 0 {
-                            b_to_amount = a.from_amount;
-                            // Verify that we can complete the partial order with natural number tokens remaining.
-                            if ((BigUint::from(b_to_amount) * BigUint::from(b.from_amount))
-                                % BigUint::from(b.to_amount))
-                                != BigUint::from(0u32)
-                            {
-                                continue;
-                            }
-                        }
-                    }
-                }
+            let mut a_to_amount = 0u16.into();
+            let mut b_to_amount = 0u16.into();
+            let (a, b) = m;
+            // Check if some orders can be completed in their entirety.
+            if b.fromAmount >= a.toAmount {
+                a_to_amount = a.toAmount.clone();
+            }
+            if a.fromAmount >= b.toAmount {
+                b_to_amount = b.toAmount.clone();
+            }
+            // Check if some orders can be completed partially.
+            if check_orders(a.clone(), b.clone(), &mut a_to_amount, b_to_amount.clone()) {
+                continue;
             }
-            if a_to_amount > 0 && b_to_amount > 0 {
-                self.process_trade(m.0, m.1, a_to_amount, b_to_amount);
+            if check_orders(b.clone(), a.clone(), &mut b_to_amount, a_to_amount.clone()) {
+                continue;
+            }
+
+            if a_to_amount > 0u16 && b_to_amount > 0u16 {
+                self.process_trade(a.id, b.id, a_to_amount, b_to_amount)?;
             }
         }
+
+        Ok(())
     }
 
-    fn process_trade(&mut self, a: OrderId, b: OrderId, a_to_amount: u128, b_to_amount: u128) {
+    fn process_trade(
+        &mut self,
+        a: OrderId,
+        b: OrderId,
+        a_to_amount: Nat,
```

Comment:
Just for my edification: what is the advantage of Nat over BigUint?

Reply:
More-so, security was worried about conversion to `u128` during calls overflowing. I've removed this conversion altogether, and while making this change I used `Nat` for inputs, outputs, and internal types just to keep it consistent. `Nat` is backed with `BigUint` anyway.